### PR TITLE
Fetch provided digest once per benchmark

### LINKED ISSUE #13578: SHA performance test
When I compared openssl 3.0 and openssl 1.1.1f, I found that the performance of alpha9 was lower than that of alpha5.
I am also comparing with OpenSSL 1.1.1f. I wonder if this can explain the problem.

All numbers are Percentage Diff (%)(alpha5-1.1.1f/1.1.1f)
alpha5 | 16 bytes | 64 bytes | 256 bytes | 1024 bytes | 8192 bytes | 16384 bytes
-- | -- | -- | -- | -- | -- | --
sha1 | -74.24694646 | -67.83789878 | -56.77331253 | -20.98058948 | -4.089929654 |-2.746844537
sha256 |  -68.18784269 | -59.0421308 | -42.57797189 | -20.2773497 | 2.462470535 | -14.52529967
sha512 |  -56.98251239 | -56.75811124 | -44.69438411 | -19.60932482 | 2.45562681 | -33.7214586

All numbers are Percentage Diff (%)(alpha9-1.1.1f/1.1.1f)
alpha9 | 16 bytes | 64 bytes | 256 bytes | 1024 bytes | 8192 bytes | 16384 bytes
-- | -- | -- | -- | -- | -- | --
sha1 |  -74.93675632 | -68.51181924 |  -55.35508324 |  -35.25122672 | -5.811513632 |  -2.937482437
sha256 | -93.95017321 |  -93.97662812 |  -88.74451967 |  -77.98002727 | -51.3641645 |  -46.51934182
sha512 |  -89.04846118 |  -93.23677374 | -84.93249856 | -76.25703978 |  -24.29898999 |  -37.92422949

using `openssl speed -evp` ,Five tests, and the results are almost the same.

### PR DESCRIPTION
**Pull request by beldmit**
Partially fixes #13578

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: apps/speed.c [hunk 1]
```diff
@@ -1697,7 +1715,7 @@ int speed_main(int argc, char **argv)
             evp_md = NULL;
             evp_cipher = EVP_get_cipherbyname(opt_arg());
```
Comment (mattcaswell) [inline apps/speed.c:1716]: It seems like we should make a similar change here too.

Comment (beldmit) [reply apps/speed.c:1716]: If you mean evp_cipher, I agree with you. But I want to fix some more performance issues for digests.

**Comment by beldmit** [comment]
The 2nd commit was tested against GOST (engine-provided) and works.

**Review by LiberteeFreedomMTL514 — APPROVED** [review]

**Comment by beldmit** [comment]
@mattcaswell, could you please take a look at the failing tests?

**Comment by beldmit** [comment]
The connection seem to work well without proxy even for the failing tests

**Comment by beldmit** [comment]
I've removed the commit breaking tests and implemented the cipher prefetch.

File: apps/speed.c (hunk 1 above)
Comment (beldmit) [reply apps/speed.c]: Done. Thanks!

File: apps/speed.c [hunk 2]
```diff
@@ -1695,9 +1729,9 @@ int speed_main(int argc, char **argv)
             break;
         case OPT_EVP:
             evp_md = NULL;
```
Comment (mattcaswell) [inline apps/speed.c:1731]: We should probably have a check somewhere to see if "-evp" has already been specified. Previously doing this was nonsensical, but harmless:
````
openssl speed -evp aes-128-cbc -evp aes-256-cbc
````
You can only specify "-evp" once, but if you do specify it twice the first instance is ignored. Now though, with the changes in this PR,  there will be a mem leak in this case.

File: apps/speed.c [hunk 3]
```diff
@@ -4091,6 +4125,10 @@ int speed_main(int argc, char **argv)
     }
     OPENSSL_free(loopargs);
     release_engine(e);
+    if (evp_md && fetched_alg)
+        EVP_MD_free(evp_md);
+    if (evp_cipher && fetched_alg)
+        EVP_CIPHER_free(evp_cipher);
```
Comment (mattcaswell) [inline apps/speed.c:4131]: Perhaps this is better:

````
if (fetched_alg) {
    EVP_MD_free(evp_md);
    EVP_CIPHER_free(evp_cipher);
}
````

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: apps/speed.c (hunk 2 above)
Comment (beldmit) [reply apps/speed.c:1735]: Done. Thanks!

File: apps/speed.c (hunk 3 above)
Comment (beldmit) [reply apps/speed.c:4134]: Done. Thanks!

File: apps/speed.c [hunk 4]
```diff
@@ -1491,6 +1493,38 @@ static int run_benchmark(int async_jobs,
...
+    }
+
+    return (EVP_MD *)EVP_get_digestbyname(name);
+}
+
+static EVP_CIPHER *obtain_cipher(const char *name)
+{
+    EVP_CIPHER *cipher = NULL;
+
+    /* Look through providers' digests */
```
Comment (mattcaswell) [inline apps/speed.c:1516]: s/digests/ciphers/

**Review by mattcaswell — APPROVED** [review]
Approved subject to the nit being fixed

File: apps/speed.c (hunk 4 above)
Comment (beldmit) [reply apps/speed.c]: Fixed and reworded the commit message. Could you please formally reapprove?

**Review by mattcaswell — APPROVED** [review]
Reconfirm

**Comment by beldmit** [comment]
Merged. Thanks!

**Pull request closed without merging**